### DISCUSSION TITLE
Optimistic updates for Relay Connections

### DISCUSSION THREAD

**Discussion by Kingdutch** [urql-graphql/urql#1232] [Q&A] [answered]
I have a `Conversation` type with a `messages` field that is typed `ConversationMessagesConnection` which in turn contains `pageInfo`, `nodes`, and `edges` (implementing the Relay specification). Graphcache is set-up to use `relayPagination` for the field. I'm trying to set-up the Graphcache with optimistic updates, but I'm unsure about my data requirements from a mutation.

I understand how I can manipulate the cache in `optimistic` or `update` methods on the cache. I could create a placeholder value for the cursor, but I'm worried this might cause invalid queries to the back-end in some cases (although I read it doesn't update with pending mutations).

Additionally, when my mutations returns a `Message` entity (which is the `node` in the `ConversationMessageEdge` used by the connection) then I still wouldn't have a cursor. I could change the mutation to return different/additional data, but I'm not sure what I'd need. Additionally, cursors are dependent on the sorting key for the list they're in to ensure they still work when data is deleted/updated.

Previous discussions that I could find here dealt mostly with simple pagination lists rather than Relay spec based pagination. Any help is appreciated :D 

EDIT to add keywords for future searchers: The same questions above could be applied to results for a subscription that needs to update a list using Relay pagination. I suspect the solution to both problems is (nearly) the same.

**Comment by kitten** [comment, marked as answer]
> I could create a placeholder value for the cursor

I believe it may be better _not_ to update the cursor or `pageInfo` at all. What has previously worked for others and in Relay for instance, afaik, is to append only the edge or item to the last page (or any appropriate one). All queries regarding the optimistic update will be blocked anyway while the mutation is ongoing, so there's no risk of "losing" the added optimistic item, and there's also no point of optimistically updating the cursor. In fact, since `optimistic` in the best case scenario only _returns data_ and doesn't use cache updates directly and `updates` _adds_ the new item to a list (regardless of whether it's optimistic or not) this should be indifferent logic in either case.

> Additionally, cursors are dependent on the sorting key for the list they're in to ensure they still work when data is deleted/updated.

So on this point and the last, cursors are basically state that is passed back to the server. Because optimistic data never passes back to the server it won't need to update. The only thing you have to be concerned about is when it's possible to pass _invalid_ cursors to the server. So the right thing is not to alter the cursor just because you don't want to queue up another query with an optimistically (read: client-side) created cursor. This is also not an issue because even if your cursor is off by one or two added items, those items will still be returned by the query either way when that next page is fetched.

So the thing here to remember is:

- From your end: Make sure not to trigger queries/mutations with variables that you created optimistically (which never becomes an issue unless you specifically pass "optimistically created IDs" to GraphQL)
- Adding things to a Relay-paginated list is always solid because even when the cursor is "behind" that actually gives refetches and subsequent pages a chance to see the added items _again_
- The `relayPagination` helper does deduplication, so even if it sees the items multiple times (e.g. you added it to the current list in the update after a mutation completed, so non-optimistic, and the next page contains it again) it will only show once
- Queries that can override optimistic updates are blocked from being sent. Graphcache has strong protections against optimistic updates accidentally becoming "committed" and against them disappearing unexpectedly.

**Reply by Kingdutch** [reply]
Scroll to the bottom for an insight I had while typing this.

Thanks for the very clear explanation! It's become a lot clearer what the different cache update functions are and how they relate to eachother. I've implemented the following optimistic update which seems to generate the data I expect to get back from the mutation and this arrives properly in the `updates.Mutation` entry for the same field. However, when I try to update data using `cache.updateQuery` the `data` in the update callback is `null`. Additionally if I try to read the fields for the conversation for which I want to mutate the messages entries then I also receive an empty array.

The query that's used for the `updateQuery` in the below example is the same as the query that's used in the `Conversation` detail view which fetches all the data I want to update (and contains the component triggering the mutation). The only difference is that the query itself is `query conversation_details($id: ID!, $last_messages: Int!) {` where `$last_messages` can be changed in the component instead of being static at `1` and the `sender`, `sent`, `content` fields (under `conversation.messages.edges.node`) are contained in a `Chat_Message`, which in the original query is under `conversation.messages.edges`.

```js

    // Implement optimistic updates for things like conversation creation or message sending.
    optimistic: {
      sendUserMessage: (variables, cache, info) => {
        const currentUser = cache.readQuery({
          query: `query {
            currentUser {
              id
              displayName
            }
          }`
        })?.currentUser ?? null;
        return {
          __typename: 'SendMessagePayload',
          clientMutationId: variables.input.clientMutationId,
          errors: null,
          message: {
            __typename: "Message",
            id: "optimistic-" + variables.input.clientMutationId,
            conversation: {
              __typename: "Conversation",
              id: variables.input.conversation,
            },
            sender: {
              __typename: "User",
              id: currentUser?.id,
              displayName: currentUser?.displayName,
            },
            sent: {
              __typename: "DateTime",
              timestamp: Date.now(),
            },
            content: {
              __typename: 'MediaMessageContent',
              text: variables.input.contents.text
            },
          }
        }
      }
    },
    // Ensure mutation and subscription results properly update our cache.
    // This completes (or rejects) optimistic updates and ensures subscription results are
    // added to overview lists that a user may be viewing.
    updates: {
      Mutation: {
        // Look into updateQuery but also reduce the query size by requesting fewer fields.
        sendUserMessage: (result, args, cache, info) => {
          const conversationFields = cache.inspectFields({ __typename: 'Conversation', id: args.input.id });
          console.log(conversationFields);
          // Update any conversations that may have this message.
          cache.updateQuery({
            query: `
              query ($id: ID!) {
                __typename
                conversation(id: $id) {
                  __typename
                  id
                  messages(last: 1) {
                    __typename
                    edges {
                      __typename
                      node {
                        __typename
                        id
                        sender {
                          __typename
                          id
                          displayName
                        }
                        sent {
                          __typename
                          timestamp
                        }
                        content {
                          ... on MediaMessageContent {
                            __typename
                            text
                          }
                        }
                      }
                    }
                  }
                }
              }
            `,
            variables: {
              id: args.input.conversation
            }
          }, data => {
            console.log("update", data);
            const { sendUserMessage } = result;
            data.conversation.messages.edges[0].push({
              node: sendUserMessage,
            });
            return data;
          })
        }
      }
    }
```
------

While typing I noticed that `messages.nodes` was not the same as `messages.edges.node` which I thought could explain the missing result in the `updateQuery`. So I could fix this by **making sure the fetched pagination from a Connection matches between the cache read and the query populating the cache** (emphasis to make sure future people read it and so that I remember it).

This leaves me with two questions:

1. Why, with the updated query filling the cache is `inspectFields` still returning an empty array? It feels like the easier method to update all message listings rather than creating the right queries everywhere.
2. Is there a way I can tell the cache that what's in `nodes` could also be in `edges` (actually, vice versa because I can understand that from nodes -> edges I'd be missing information, but that's not the case from edges -> nodes). So that both can be updated.

(Bonus question, is there anything else with my current approach that you'd recommend against? I feel like I've had to add `__typename` all over the place too, so I'm wondering if I can omit that somewhere)

------

Edit after the above:

The `updateQuery` only returns data if I use `messages.nodes` in both places, but not if I use `messages.edges.node` in both places. With the latter case, data is once again `null`. (`messages` returns a `ConversationMessageConnection` that has `edges: MessageEdge!`, `nodes: [Message!]!`, `pageInfo: PageInfo!`. `MessageEdge` has `cursor: Cursor!` and `node: Message!`)

------
Further update:

After changing both queries to use `nodes` for now, I've been able to append the data to the right place in a populated data entry in an update query. Unfortunately, this doesn't cause my React component to rerender with the new data. (it also won't allow for pagination, since the set-up with the cursor doesn't work).

**Reply by Kingdutch** [reply]
## Solution
With a couple of hours more digging I was able to figure out that my components weren't re-rendering because the queries used in my components were sufficiently different from the query in my cache. Most notably my component was pulling in `messages(last: $last_messages)` with `20` last messages, whereas my cache query only loaded `messages(last: 1)`. Changing my cache query to also have `messages(last: 20)` ensures that optimistic updates and updates from subscriptions worked as expected.

Testing has shown that changing `last: ` to `19` or to `21` (basically something other than what's used in the component) will not work. This makes sense if the cache is not specifically aware of the pagination and thus treats different arguments as different cache sets. This means that the cache queries need to be kept in sync with the component queries.

## Working snippet
The following snippets successfully adds the mentioned update to something using relay pagination. This snippet performs an optimistic update, then updates the cache once the mutation is complete. It'll also update the cache when it receives related entities from a subscription. Due to a server misimplementation the mutated entities also come through the subscription so some deduplication is needed there. You could implement an update instead of an ignore strategy.

The cache configuration
```js
    // Implement optimistic updates for things like conversation creation or message sending.
    optimistic: {
      sendUserMessage: (variables, cache, info) => {
        const currentUser = cache.readQuery({
          query: `query {
            currentUser {
              id
              displayName
            }
          }`
        })?.currentUser ?? null;
        return {
          __typename: 'SendMessagePayload',
          errors: null,
          message: {
            __typename: "Message",
            id: "optimistic-" + variables.input.clientMutationId,
            conversation: {
              __typename: "Conversation",
              id: variables.input.conversation,
            },
            sender: {
              __typename: "User",
              id: currentUser?.id,
              displayName: currentUser?.displayName,
            },
            sent: {
              __typename: "DateTime",
              timestamp: Math.floor(Date.now() / 1000),
            },
            content: {
              __typename: 'MediaMessageContent',
              text: variables.input.contents.text
            },
          }
        }
      }
    },
    // Ensure mutation and subscription results properly update our cache.
    // This completes (or rejects) optimistic updates and ensures subscription results are
    // added to overview lists that a user may be viewing.
    updates: {
      Mutation: {
        sendUserMessage: (result, args, cache, info) => {
          console.log("sendUserMessage", result);
          const message = result.sendUserMessage.message;
          // Our optimistic update doesn't include a clientMutationId but our result does,
          // so we use it to remove the optimistically added message.
          if (typeof result.clientMutationId !== "undefined") {
            cache.invalidate({__typename: 'Message', id: 'optimistic-' + result.clientMutationId});
          }
          // Update any conversations that may have this message.
          cache.updateQuery({
            query: MessageUpdateQuery,
            variables: {
              id: args.input.conversation
            }
          }, data => {
            console.log("update", data);
            if (message === null) {
              // TODO: Implement some kind of error state.
              return null;
            }

            // TODO: implement support for edges
            // Append the new (optimistic or saved) message.
            if (data.conversation.messages.nodes === null) {
              data.conversation.messages.nodes = [];
            }
            data.conversation.messages.nodes.push(message);
            return data;
          })
        }

      },
      Subscription: {
        messageReceived: (result, args, cache, info) => {
          console.log("messageReceived", result);
          const message = result.messageReceived;
          // Update any conversations that may have this message.
          cache.updateQuery({
            query: MessageUpdateQuery,
            variables: {
              id: result.messageReceived.conversation.id
            }
          }, data => {
            console.log("update", data);
            if (message === null) {
              // TODO: Implement some kind of error state.
              return null;
            }

            // TODO: implement support for edges
            // If this message already exists, update it.
            // We just ignore existing messages. We don't expect changes here.
            if (data.conversation.messages.nodes.findIndex(m => m.id === message.id) !== -1) {
              return null;
            }

            // Otherwise append it.
            if (data.conversation.messages.nodes === null) {
              data.conversation.messages.nodes = [];
            }
            data.conversation.messages.nodes.push(message);

              return data;
          })
        }


      }
    }
```

The referred query in those updates.
```js
 const MessageUpdateQuery = `
 query ($id: ID!) {
  __typename
  conversation(id: $id) {
    __typename
    id
    messages(last: 20) {
      __typename
      nodes {
        __typename
        id
        sender {
          __typename
          id
          displayName
        }
        sent {
          __typename
          timestamp
        }
        content {
          ... on MediaMessageContent {
            __typename
            text
          }
        }
      }
    }
  }
}
 `;

```

## Remaining questions

This still leaves me with a couple of open questions.

1. Can I make the cache aware of the relationship between edges and nodes?
2. Why was inspectFields not returning any data?
3. What is the best way to organise this cache logic in a project to ensure that the components and the cache don't go out of sync?

**Reply by kitten** [reply]
> Can I make the cache aware of the relationship between edges and nodes?

Yes and no; for cache updates like `cache.updateQuery` it's important to note that you have to be exact since no resolvers are used unless you use `inspectFields` to dynamically find what you need to update.

If you use `edges` and `nodes` then you need to modify the resolver yourself. You can for instance wrap around a `relayPagination()` resolver and map all `edges` results to `nodes`. While we support both however try not to mix them. Remember that your cache is storing pages of results. If you have some that use `edges` and some that use `nodes` then you will likely get confused. The former is Relay-spec compliant, the latter is not. But we don't expect a schema (or an app at least) to use both.

> Why was inspectFields not returning any data?

I can't see your code snippet for it, so I can't tell you. But check the arguments you're passing to it agains the docs.

> What is the best way to organise this cache logic in a project to ensure that the components and the cache don't go out of sync?

Consumers of urql's cache data and the cache never go out of sync. So I assume you mean what you should do about specific `updateQuery` arguments and the likes? When dealing with variable arguments for pagination fields we always recommend `cache.inspectFields` to keep the logic generic, since that removes the need to hardcode inputs across the app.

**Reply by Kingdutch** [reply]
> If you have some that use edges and some that use nodes then you will likely get confused. 

Ha it seems that's what I'm running into already.

> The former is Relay-spec compliant, the latter is not. But we don't expect a schema (or an app at least) to use both.

Okay, I did not know that. I found [information on the web](https://medium.com/javascript-in-plain-english/graphql-pagination-using-edges-vs-nodes-in-connections-f2ddb8edffa0) that offered `nodes` as useful shortcut in places where you don't need cursors and [GitHub also uses them](https://docs.github.com/en/free-pro-team@latest/graphql/reference/objects#securityadvisoryconnection). 

I now understand the problems that causes, so I'll have to stick to using edges. I'll start the discussion internally on whether we keep `nodes` in our public schema or remove it there as well :) 

> I can't see your code snippet for it, so I can't tell you. But check the arguments you're passing to it agains the docs.

Ah I see I forgot that snippet, sorry! I'll run with the new information you've given me and try again and bring back the results with some code :D 

> So I assume you mean what you should do about specific updateQuery arguments and the likes?

Yes! I had no doubt about the proper functioning of the cache. I was indeed mostly wondering about keeping the arguments for queries in my components up to date with the logic in the cache updaters. Sounds like making sure I properly implement `cache.inspectFields` will solve that problem, awesome!

**Reply by Kingdutch** [reply]
## Eureka
Okay, whew that took me a time to figure out. However, I've learned the following, which I didn't quite understand from the documentation.

Although the `cache.inspectFields` works on partial entities, it does not return a result if there could be more than a single entity that matches. It will only return results if it can match to a specific entity of concrete types.

The example in the documentation is somewhat misleading because it uses `Query` as type, which is probably the only of three types (Query, Mutation, Subscription) that doesn't require an ID.

Take the following 7 inspectField examples that I've found as example. The queries executed in the application (leaf and id/__typename fields omitted, but present) when the code is executed in a `updates.mutation.sendUserMessage` function on the cache are as follows: first a query for `currentUser { conversations(last: 10) { messages(last: 1) { ... } } }`, then a query for `conversation(id: ...) { messages(last: 20) { user: sender { ....} ... } }`, then the mutation.

```javascript
console.log(cache.inspectFields({ __typename: 'Query' }));  // 1
console.log(cache.inspectFields({ __typename: 'User' }));  // 2
console.log(cache.inspectFields({ __typename: 'Conversation' })); // 3
console.log(cache.inspectFields({ __typename: 'Conversation', id: args.input.conversation }));  // 4
console.log(cache.inspectFields({ __typename: 'PrivateConversation' });  // 5
console.log(cache.inspectFields({ __typename: 'PrivateConversation', id: args.input.conversation  })); // 6
console.log(cache.inspectFields({ __typename: 'Message' })); // 7
```

inspectField 1 returns an array with 2 fields (`currentUser` and `conversation(id: ...)`). inspectField 2, 3, 4, 5, 7 all return an empty array. Only inspectField 6 gave me the output that I was looking for that included the two usages of the `messages` field independent of their arguments, allowing me to implement robust update logic.

It wasn't clear for me from reading [the example](https://formidable.com/open-source/urql/docs/graphcache/custom-updates/#cacheinspectfields) that I had to find a specific entity, not just all entities of a type. That distinction only clicked when I re-read the first line of [the API docs](https://formidable.com/open-source/urql/docs/api/graphcache/#inspectfields) for an uncounted time. Although the example there also still uses `Query` which made it initially unclear (since my mind never treated the result of `Query` as a single entity, although of course, it is).

With what I know now, almost all of this makes sense. My troubles were exacerbated slightly by the fact that I had tried inspectField 4 but never tried inspectField 6. I did not know that `inspectField` had to be called on a concrete type, even though the field I'm interested in is part of the `Conversation` interface, so it'd be present for all implemented types.

## Proposed documentation change

Perhaps we can amend the example in the documentation in some way to drive home the fact that you're always looking for a single entity and amend the API docs to not use `Query` but a different type for its example, to help enforce the first line of that section of the documentation.

## Requested feature change
Could the `inspectField` method be made to work with interfaces as the type (when schema awareness is enabled, which is the case in my trials).

If that's not possible due to a technical limitation, perhaps we can add that to the docs too.


-----

Thanks for all the information! I hope my journey and gained insights can save someone else a few days of trial and error ^^

**Reply by kitten** [reply]
Just as a tip 😅 try to keep your writing shorter. You can assume that readers have prior knowledge especially, since core contributors are the only ones likely to read in detail up until here 😀

Anyway, we can make sure that the guide is fixed up. I believe I'll just change this line:

> and is able to return all fields that the cache has seen on a given type.

It doesn't use the usual phrasing, I.e. "entity", and is hence confusing.

**Reply by Kingdutch** [reply]
Okay I do run into another hurdle now that I have the fields that I want to update.

The data requested on the fields is different so the data in the cache may be different too. I can't seem to find a reliable query to update both returned fields:
```json
[
  {
    "fieldKey": "messages({\"last\":1})",
    "fieldName": "messages",
    "arguments": {
      "last": 1
    }
  },
  {
    "fieldKey": "messages({\"last\":20})",
    "fieldName": "messages",
    "arguments": {
      "last": 20
    }
  }
]
```

If I use my original query `MessageUpdateQuery` it'll work for the `last 20` field but the returned data will be `null` for `last: 1` because the latter query initially doesn't need as much data, so not all requested fields are present.

If I request only the message `id` for both instances, then I get a result in both cases but it seems that I wipe out the cache for the other messages that are returned by the `updateQuery`, since I'm returning an object with far fewer fields, for all entities except for the new one.

I don't see a way to properly request the fields that were included in the query for the field. Even using more recursive `inspectFields` and `resolve` queries is problematic because I wouldn't be able to distinguish between a field value and an entity relationship.

This is the problematic code that'll either throw away a lot of message data if executed only for `last: 20` or will seemingly trash the cache in case of `last: 1` returning no data for the subsequent `last: 20` query.

```javascript
messageQueries.forEach(({ arguments: fieldArgs }) => {
            console.log(args.input.conversation, fieldArgs);

            cache.updateQuery({
              query: `query ($id: ID!, $last: Int!) {
                conversation(id: $id) {
                  messages(last: $last) {
                    nodes {
                      id
                    }
                  }
                }
              }`,
              variables: {
                id: args.input.conversation,
                last: fieldArgs.last,
              }
            }, data => {
              console.log(data);
              // TODO: implement support for edges
              // If there were no results previously then we have to initialise a new connection.
              if (data.conversation.messages.nodes === null) {
                data.conversation.messages.nodes = [];
              }
              // In case it's the `last: 1` we replace it since the query is only expecting one result.
              if (fieldArgs.last === 1) {
                data.conversation.messages.nodes[0] = message;
              }
              // Otherwise we append the new message in our cache.
              else {
                data.conversation.messages.nodes.push(message);
              }

              return data;
            });
          });
```

**Reply by kitten** [reply]
You'll want to append the message to each ending list that doesn't have a `before` cursor then and not replace any nodes. The length won't matter as long as your UI just display them, especially since the lists will be merged by `relayPagination`

**Reply by Kingdutch** [reply]
> Just as a tip 😅 try to keep your writing shorter. You can assume that readers have prior knowledge especially, since core contributors are the only ones likely to read in detail up until here 😀

Fair enough, I'm just trying to make sure that it's clear what I'm doing (and making sure that if I, as non-contributor) come back in 3 months I can also still figure it out 😊  I'll try to tighten up my writing here and leave the longer writings for blog posts.

> You'll want to append the message to each ending list that doesn't have a before cursor then and not replace any nodes. The length won't matter as long as your UI just display them, especially since the lists will be merged by relayPagination

Okay, that's what my last snippet did. However, it turned out that because my `updateQuery` didn't request the `conversation.id` field the cache wasn't able to reconcile the results. I managed to find the warning for that and it looks like everything is now working as expected! 🎉
